tests/unit/lms/services/grouping_test.py

```diff
@@ -29,8 +30,35 @@ def test_upsert_with_parents_inserts(self, svc, db_session):
 
         assert db_session.query(CanvasGroup).one() == test_grouping
 
+    def test_you_cant_upsert_a_grouping_whose_parent_has_a_different_application_instance(
+        self, svc, application_instance_service, db_session
+    ):
+        parent = factories.Course()
+        # GroupingService uses ApplicationInstanceService.get_current() as the
+        # application_instance for the new grouping it inserts.
+        # Here the parent course will have a different application_instance,
+        # which will trigger an IntegrityError when we flush the DB.
+        assert (
+            parent.application_instance
+            != application_instance_service.get_current.return_value
+        )
+
+        svc.upsert_with_parent(
+            tool_consumer_instance_guid="tool_consumer_instance_guid",
+            lms_id="lms_id",
+            lms_name="lms_name",
+            type_=Grouping.Type.CANVAS_GROUP,
```

Comment:
Do any of these four values matter? It might be nice to try and explain they don't (if they don't) so the important ones stand out. Maybe with values like "any_string" or `Any` objects or something. Whatever we can get away with.

Reply:
They don't matter but I can't think of a great solution:
 
* You can't use something like `Any` because it actually has to be a string because it's actually going to go into the DB
* I don't think hard-coded strings like `"tool_consumer_instance_guid"` or even `"any_string"` really get across the message that these values don't matter to this test. And they're visually noisy: they obliterate the `parent=parent` argument that actually does matter
* I thought of maybe a `random_string()` function but it's still visually noisy and you don't want actual randomness in tests
* There's also the `type_=Grouping.Type.CANVAS_GROUP` which is not a string. This can be any `Grouping.Type.*` except `Grouping.Type.CANVAS_GROUP` because course's don't have parents

factory_boy does this so well when creating model objects for us: allowing each test to specify only the values that it cares about, and having appropriate values be generated for all the other attributes.

We _could_ actually write a factory for generating the args to the `upsert_with_parent()` method:

```diff
diff --git a/tests/unit/lms/services/grouping_test.py b/tests/unit/lms/services/grouping_test.py
index 72e10a35..4c2c35fe 100644
--- a/tests/unit/lms/services/grouping_test.py
+++ b/tests/unit/lms/services/grouping_test.py
@@ -1,5 +1,6 @@
 from unittest.mock import sentinel
 
+import factory as factory_boy
 import pytest
 from sqlalchemy.exc import IntegrityError
 
@@ -10,6 +11,16 @@ from tests import factories
 pytestmark = pytest.mark.usefixtures("course_service", "application_instance_service")
 
 
+upsert_with_parent_args = factory_boy.make_factory(
+    dict,
+    # In the real version you'd randomise these hard-coded strings.
+    tool_consumer_instance_guid="tool_consumer_instance_guid",
+    lms_id="lms_id",
+    lms_name="lms_name",
+    parent=None,
+    type_=Grouping.Type.CANVAS_GROUP,
+)
+
+
 class TestGroupingService:
     CONTEXT_ID = "context_id"
     TOOL_CONSUMER_INSTANCE_GUID = "t_c_i_guid"
@@ -43,13 +54,7 @@ class TestGroupingService:
             != application_instance_service.get_current.return_value
         )
 
-        svc.upsert_with_parent(
-            tool_consumer_instance_guid="tool_consumer_instance_guid",
-            lms_id="lms_id",
-            lms_name="lms_name",
-            type_=Grouping.Type.CANVAS_GROUP,
-            parent=parent,
-        )
+        svc.upsert_with_parent(**upsert_with_parent_args(parent=parent))
 
         with pytest.raises(
             IntegrityError,
```

Reply:
I'm gonna merge the PR as-is and explore this in a separate PR